source/extensions/filters/udp/dns_filter/dns_parser.h

```diff
@@ -0,0 +1,192 @@
+#pragma once
+
+#include "envoy/buffer/buffer.h"
+#include "envoy/network/address.h"
+#include "envoy/network/listener.h"
+
+#include "common/buffer/buffer_impl.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace UdpFilters {
+namespace DnsFilter {
+
+// These flags have been verified with dig. The flag order does not match the RFC, but takes byte
+// ordering into account so that serialization does not need bitwise operations
+PACKED_STRUCT(struct DnsHeaderFlags {
+  unsigned rcode : 4;  // return code
+  unsigned cd : 1;     // checking disabled
+  unsigned ad : 1;     // authenticated data
+  unsigned z : 1;      // z - bit (must be zero in queries per RFC1035)
+  unsigned ra : 1;     // recursion available
+  unsigned rd : 1;     // recursion desired
+  unsigned tc : 1;     // truncated response
+  unsigned aa : 1;     // authoritative answer
+  unsigned opcode : 4; // operation code
+  unsigned qr : 1;     // query or response
+});
+
+/**
+ * Structure representing the DNS header as it appears in a packet
+ * See https://www.ietf.org/rfc/rfc1035.txt for more details
+ */
+PACKED_STRUCT(struct DnsHeader {
+  uint16_t id;
+  struct DnsHeaderFlags flags;
+  uint16_t questions;
+  uint16_t answers;
+  uint16_t authority_rrs;
+  uint16_t additional_rrs;
+});
+
+enum DnsRecordClass { IN = 1 };
+enum DnsRecordType { A = 1, AAAA = 28 };
+enum DnsResponseCode { NoError, FormatError, ServerFailure, NameError, NotImplemented };
+
+/**
+ * BaseDnsRecord contains the fields and functions common to both query and answer records.
+ */
+class BaseDnsRecord {
+public:
+  BaseDnsRecord(const uint16_t id, const std::string& rec_name, const uint16_t rec_type,
+                const uint16_t rec_class)
+      : id_(id), name_(rec_name), type_(rec_type), class_(rec_class){};
+
+  virtual ~BaseDnsRecord() = default;
+  void serializeName();
+  virtual void serialize(Buffer::OwnedImpl& output) PURE;
+
+  const uint16_t id_;
+  const std::string name_;
+  const uint16_t type_;
+  const uint16_t class_;
+
+protected:
+  Buffer::OwnedImpl buffer_;
+};
+
+/**
+ * DnsQueryRecord represents a query record parsed from a DNS request from a client. Each record
+ * contains the ID, domain requested and the flags dictating the type of record that is sought.
+ */
+class DnsQueryRecord : public BaseDnsRecord {
+
+public:
+  DnsQueryRecord(const uint16_t id, const std::string& rec_name, const uint16_t rec_type,
+                 const uint16_t rec_class)
+      : BaseDnsRecord(id, rec_name, rec_type, rec_class) {}
+
+  ~DnsQueryRecord() override = default;
+  void serialize(Buffer::OwnedImpl& output) override;
+};
+
+using DnsQueryRecordPtr = std::shared_ptr<DnsQueryRecord>;
+using DnsQueryMap = absl::flat_hash_map<uint16_t, std::list<DnsQueryRecordPtr>>;
```

Comment:
Would an unordered_multimap be a better choice here?

Reply:
Possibly.  I had that thought as well, but went with this implementation.  Let me see what I can do for this.